pandas/_libs/tslibs/parsing.pyx

```diff
@@ -904,6 +907,53 @@ def guess_datetime_format(
     #  that any user-provided function will not either.
     tokens = dt_str_split(dt_str)
 
+    # Normalize offset part of tokens.
+    # There are multiple formats for the timezone offset.
+    # To pass the comparison condition between the output of `strftime` and
+    # joined tokens, which is carried out at the final step of the function,
+    # the offset part of the tokens must match the '%z' format like '+0900'
+    # instead of ‘+09:00’.
+    if (parsed_datetime.tzinfo is not None):
+        if (len(tokens) > 0) and (tokens[-1] == 'Z'):
+            # the last 'Z' means zero offset
+            tokens[-1] = '+0000'
+        else:
+            # If the input string has a timezone offset like '+0900',
+            # the offset is separated into two tokens, ex. ['+', '0900’].
+            # This separation will prevent subsequent processing
+            # from correctly parsing the time zone format.
+            # So in addition to the format nomalization, we rejoin them here.
+            if (
+                (len(tokens) > 3)
+                and tokens[-1].isdigit()
+                and (tokens[-2] == ':')
```

Comment:
To be safe, I think this should be any non-digit separator just in case some format using something different than a colon

Reply:
I skipped the check about the separator string. So no assumptions are made about the separator type. 
If you think that the `not tokens[-2].isdigit()` is better, please let me know.